src/components/molecules/EmojiPicker/EmojiPicker.tsx

```diff
@@ -0,0 +1,27 @@
+import React, { Suspense, lazy } from "react";
+
+import { EmojiData } from "emoji-mart";
+
+import { tracePromise } from "utils/performance";
+
+import { Loading } from "components/molecules/Loading";
+
+import "emoji-mart/css/emoji-mart.css";
+
+const Picker = lazy(() =>
+  tracePromise("EmojiPicker::lazy-import::emoji-mart", () =>
+    import("emoji-mart").then(({ Picker }) => ({
+      default: Picker,
+    }))
+  )
+);
```

Comment:
Is it possible to transform the lazy load logic into a HOC or hook and just reuse it?

Reply:
Probably OOS of this PR.

I tried playing around with extracting a util/similar for it but was running into some painful if not impossible aspects with the typing.

My ideal would be to pass in just the name of the module and the thing we want to extract from it, but that didn’t seem to work.

It might be possible to provide a () => import(‘foo’) and then a key to extract from that.. shall give it another try this morning and see how that turns out.

Reply:
Well.. here is my attempt..

```typescript
import { ComponentType, LazyExoticComponent, lazy } from "react";

import { tracePromise } from "./performance";

export const lazyImportComponent = <
  T extends Record<string, ComponentType<U>>,
  U
>(
  lazyImport: () => Promise<T>,
  component: keyof T,
  traceName: string
): LazyExoticComponent<T[keyof T]> =>
  lazy(() =>
    tracePromise(traceName, () =>
      lazyImport().then((module) => {
        if (!(component in module))
          throw new Error("component not found in module");

        return { default: module[component] };
      })
    )
  );
```

Which should be usable like this:

```typescript
import { lazyImportComponent } from "utils/import";

const Picker = lazyImportComponent(() => import("emoji-mart"), "Picker", "TODO");
```

Except TypeScript says no.. so if you can figure out solving this.. we can do it:
![image](https://user-images.githubusercontent.com/753891/124588946-c1a4ce80-de9c-11eb-9693-3b77c0003afc.png)

I believe the issue is in this part of it most likely:

```typescript
T extends Record<string, ComponentType<U>>,
```

Reply:
Slack thread for figuring it out in future. Calling it OOS of this PR.

- https://doesitsparkle.slack.com/archives/C01ER4M61TL/p1625569313087900

Reply:
> Except TypeScript says no.. so if you can figure out solving this.. we can do it:

I can give it a try tonight, if I have some free time.

Reply:
> Is it possible to transform the lazy load logic into a HOC or hook and just reuse it?

The first time I tried it, it worked. I committed it to a branch. Afterwards, I merged it into another branch and suddenly the same code was throwing compiler errors.

Last time I had seen something similar was an Angular router (same problem basically, lazy loading) that did static analysis of the code, verbatim, regardless the dynamic result is correct.

Agree on OOS

Reply:
Theoretically figured it out today. Will test if it works on this PR, and if so, add it in:

```typescript
// TODO: move this to types/utilities
// This is copied from an answer at
//   https://stackoverflow.com/questions/54520676/in-typescript-how-to-get-the-keys-of-an-object-type-whose-values-are-of-a-given
// This uses TypeScript mapped types
//   https://www.typescriptlang.org/docs/handbook/2/mapped-types.htm
//   https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-1.html#mapped-types
// The -? removes the optional property
//   https://www.typescriptlang.org/docs/handbook/2/mapped-types.html#mapping-modifiers
// It then maps to a conditional type
//   https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html
// That checks if the type of T[K] extends from the type of V
//   If it does, then it maps to K, otherwise it maps to never
// Finally it uses a lookup type
//   https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-1.html#keyof-and-lookup-types
export type KeysMatching<T, V> = {
  [K in keyof T]-?: T[K] extends V ? K : never;
}[keyof T];

export const lazyImportComponent = <
  T extends Record<string, unknown>,
  U extends KeysMatching<T, ComponentType<unknown>>
>(
  lazyImport: () => Promise<T>,
  componentName: U
): LazyExoticComponent<ComponentType<T[U]>> =>
  lazy(() =>
    lazyImport().then((module) => {
      if (!(componentName in module))
        throw new Error("component not found in module");

      return { default: module[componentName] };
    })
  );

export const lazyImportComponentWithTrace = <
  T extends Record<string, unknown>,
  U extends KeysMatching<T, ComponentType<unknown>>
>(
  lazyImport: () => Promise<T>,
  componentName: U,
  traceName: string
): LazyExoticComponent<ComponentType<T[U]>> =>
  lazyImportComponent(
    () => tracePromise(traceName, () => lazyImport()),
    componentName
  );
```

Reply:
Nope.. still having issues, not sure why, but I expect it's probably saying that there are no exports that match `ComponentType<unknown>`. So.. probably a future problem to deal with in a standalone PR:

![image](https://user-images.githubusercontent.com/753891/124697983-2b6bb980-df2b-11eb-8c54-2b7798b4097e.png)